reactor-core/src/main/java/reactor/util/retry/Retry.java

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.util.retry;
+
+import java.time.Duration;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Schedulers;
+
+import static reactor.util.retry.RetryBuilder.*;
+
+/**
+ * Functional interface to configure retries depending on a companion {@link Flux} of {@link RetrySignal},
+ * as well as builders for such {@link Flux#retryWhen(Retry)} retries} companions.
+ *
+ * @author Simon Baslé
+ */
+@FunctionalInterface
+public interface Retry {
+
+	/**
+	 * The intent of the functional {@link Retry} class is to let users configure how to react to {@link RetrySignal}
+	 * by providing the operator with a companion publisher. Any {@link org.reactivestreams.Subscriber#onNext(Object) onNext}
+	 * emitted by this publisher will trigger a retry, but if that emission is delayed compared to the original signal then
+	 * the attempt is delayed as well. This method generates the companion, out of a {@link Flux} of {@link RetrySignal},
+	 * which itself can serve as the simplest form of retry companion (indefinitely and immediately retry on any error).
+	 *
+	 * @param retrySignalCompanion the original {@link Flux} of {@link RetrySignal}, notifying of each source error that
+	 * _might_ result in a retry attempt, with context around the error and current retry cycle.
+	 * @return the actual companion to use, which might delay or limit retry attempts
+	 */
+	Publisher<?> generateCompanion(Flux<RetrySignal> retrySignalCompanion);
+
+	/**
+	 * State for a {@link Flux#retryWhen(Retry) Flux retry} or {@link reactor.core.publisher.Mono#retryWhen(Retry) Mono retry}.
+	 * A flux of states is passed to the user, which gives information about the {@link #failure()} that potentially triggers
+	 * a retry as well as two indexes: the number of errors that happened so far (and were retried) and the same number,
+	 * but only taking into account <strong>subsequent</strong> errors (see {@link #failureSubsequentIndex()}).
+	 */
+	interface RetrySignal {
+
+		/**
+		 * The ZERO BASED index number of this error (can also be read as how many retries have occurred
+		 * so far), since the source was first subscribed to.
+		 *
+		 * @return a 0-index for the error, since original subscription
+		 */
+		long failureTotalIndex();
+
+		/**
+		 * The ZERO BASED index number of this error since the beginning of the current burst of errors.
+		 * This is reset to zero whenever a retry is made that is followed by at least one
+		 * {@link org.reactivestreams.Subscriber#onNext(Object) onNext}.
+		 *
+		 * @return a 0-index for the error in the current burst of subsequent errors
+		 */
+		long failureSubsequentIndex();
+
```

Comment:
I think the names are not very self-explanatory. The Javadoc describes it well, so taking words from the Javadoc, `totalRetries` or `totalFailures` perhaps along with `totalRetriesInSequence` or `totalFailuresInSequence` (or `~InARow`).

Reply:
or maybe `totalSequential{Retries,Failures}`?